### DISCUSSION TITLE
option for conditional bias estimation within the imputation routine

### DISCUSSION THREAD

**Discussion by sheffe** [kogalur/randomForestSRC#343] [General]
I do a lot of work with RFs on data where the conditional prediction intervals/conditional biases in the tails of the label range are interesting, and where we want to end up with a locally unbiased model over the full domain of the label. This also comes up a lot in RF-based imputation, where the distributions of real values are usually much fatter-tailed than distributions of imputed values.

Handwaving a bunch here, but my three-sentence summary of a lot of literature is: (1) RFs tend to be excellent estimators of a conditional mean *except* in the tails and at significant discontinuities; (2) the conditional bias in the tails/at discontinuities is often quite stable and supports a second-step estimation of a bias model (often just a lm or some extra boosting); and (3) there are some new approaches that can estimate the bias and prediction intervals entirely internal to the primary RF model. For (3) this is the ArXiv preprint of my favorite new RF paper (https://arxiv.org/abs/1912.07435) which has an R package `forestError` (on CRAN 3 days ago) that works natively with `randomForestSRC` models. The paper has a much better lit review than I could hope to provide.

Here's an image from another issue I filed to visualize this -- the plot is from an mForest imputation on 9 features from an artificial dataset (code at the end). For each case, we're globally unbiased but with varying levels of local conditional bias in the tails. 
![impute_tail_bias](https://user-images.githubusercontent.com/3778733/72628912-4e607300-391d-11ea-9d52-4d70588a33c0.png)

It's easy to use the `forestError` package on trained rfsrc models when the model is returned, but the `impute` step returns a dataset only and the training/mForest iteration is all internal. It would be an interesting option to enable a conditional bias step within that routine. I'm not familiar enough with the `randomForestSRC` package internals to understand the complexity of a move like that, but plan to dig in further. In the meantime, I hope this reference and concept are interesting.

--- code 
```r
# remotes::install_version("randomForestSRC", "2.9.1")
# needed to install the older version to avoid impute bug at runtime of this code

library(tidyverse)
library(randomForestSRC)

test_rows <- 20000

set.seed(1234)
fake_data <- dplyr::tibble(
  a = rpois(test_rows, 500),
  b = a + rnorm(test_rows, 50, 50),
  c = b + rnorm(test_rows, 50, 50),
  d = c + rnorm(test_rows, 50, 50),
  e = d + rnorm(test_rows, 50, 50),
  f = e + rnorm(test_rows, 50, 50),
  g = f + rnorm(test_rows, 50, 50),
  h = g + rnorm(test_rows, 50, 50),
  i = h + rnorm(test_rows, 50, 50)
)

fake_data_missing <- fake_data %>%
  dplyr::mutate_if(is.numeric,
                   list(~ifelse(runif(test_rows) <= 0.1, NA, .)))

cor(fake_data)

imputed_data <- randomForestSRC::impute(
  formula = a ~ b + c + d + e + f + g + h + i,
  data = as.data.frame(fake_data_missing),
  mf.q = 0.2,
  ntree = 100,
  verbose = TRUE
)

fake_long <- fake_data %>%
  dplyr::mutate(row_n = row_number()) %>%
  tidyr::gather(key = feature, value = rawval, -row_n)

imputed_long <- imputed_data %>%
  dplyr::mutate(row_n = row_number()) %>%
  tidyr::gather(key = feature, value = imputedval, -row_n)

dplyr::full_join(fake_long, imputed_long, by = c("row_n", "feature")) %>%
  # drop out cases of exact equality
  dplyr::filter(rawval != imputedval) %>%
  ggplot(data = ., aes(x = rawval, y = imputedval)) +
  facet_wrap(~feature, scales = "free") +
  geom_abline(aes(intercept = 0, slope = 1),
              color = "darkred") +
  geom_point(size = 0.2, alpha = 0.5) +
  geom_smooth()

```

**Comment by ishwaran** [comment]
Excellent example and thanks.  We looked at this several times and have plans in the future to include utilities for analyzing performance of the imputation.
